ndcube/ndcollection.py

```diff
@@ -0,0 +1,202 @@
+import collections.abc
+
+import numpy as np
+
+from ndcube.utils.cube import convert_extra_coords_dict_to_input_format
+import ndcube.utils.collection as collection_utils
+
+__all__ = ["NDCollection"]
+
+#class NDCollection(collections.abc.Sequence):
+class NDCollection:
+    def __init__(self, data, labels=None, aligned_axes=None, dont_sanitize_aligned_axes=False):
+        """
+        A class for holding and manipulating a collection of aligned NDCube or NDCubeSequences.
+
+        Data cubes/sequences must be aligned, i.e. have the same WCS and be the same shape.
+
+        Parameters
+        ----------
+        data: sequence of `~ndcube.NDCube` or `~ndcube.NDCubeSequence`
+            The data cubes/sequences to held in the collection.
+
+        labels: sequence of `str`
+            Name of each cube/sequence. Each label must be unique and
+            there must be one per element in the data input.
+            Default is ("0", "1",...)
+
+        aligned_axes: `tuple` of `int`, `tuple` of `tuple`s of `int`
```

Comment:
better to use physical type here?

Reply:
That's a much more elegant solution.  Two things we would have to think through first:
* Could users want to align axes that aren't of the same physical type?  As it is now, the only requirement is that the aligned axes are the same length which gives maximum flexibility.
* Could users have axes with the same physical type?  We haven't addressed such a use case in ```NDCube```, but I wonder if it's worth considering here?

Reply:
Perhaps we could support both APIs?  A ```tuple``` of ```str``` is parsed as physical axis types while a ```tuple``` of ```int``` or ```tuple``` is parsed as data axis numbers.  Either way, the information should be stored under-the-hood in the way the current API is set up.

Reply:
On second thoughts, are we able to support using physical type at the moment.  We had a previous discussion that revealed that strictly speaking, ```NDCube.world_axis_physical_types``` does not always have a one-to-one correspondence with the data axes, especially in the cases of coupled axes and rotated data.  Although in most cases this is effectively true, we nonetheless cannot strictly, uniquely derive the data axes from the world_axis_physics_types.  We can only go the other way.  Therefore we need to:
1. Redefine ```NDCube.world_axis_physical_types```
2. Create a new attribute that more strictly related physical axis types to data axes, even in the case of coupled axes.
3. Not use physical types here to derive the aligned data axes.

Thoughts?